src/parser.rs

```diff
@@ -0,0 +1,555 @@
+            '&' | '<' | '>' | ';' | '|' if !meta_ignored => {
+                debug!("Token: {}", current_token);
+                if !current_token.is_empty() {
+                    tokens.push(Token::Plain(current_token.clone()));
+                    current_token.clear();
+                }
+
+                // Check for >>
+                if c == '>' && tokens.last() == Some(&Token::Meta(">".to_string())) {
+                    tokens.pop();
+                    tokens.push(Token::Meta(">>".to_string()));
+                }
+                // Push meta character as a token
+                else {
+                    tokens.push(Token::Meta(c.to_string()));
+                }
+            }
+            ' ' if !meta_ignored => {
+                debug!("Found whitespace");
+                if !current_token.is_empty() {
+                    tokens.push(Token::Plain(current_token.clone()));
+                    current_token.clear();
+                }
+            }
+            '\\' if !meta_ignored => {
+                debug!("Found backslash");
+                escaped = true;
+            }
+            '\'' | '"' if !escaped => {
+                debug!("Found quote");
+
+                // If the quote is closed, pop it from the stack
+                if quote_stack.last() == Some(&c) {
+                    match c {
+                        '\'' => {
+                            quote_stack.pop();
+                            tokens.push(Token::SingleQuoted(current_token.clone()));
+                            current_token.clear();
+                        }
+                        '"' => {
+                            quote_stack.pop();
+                            tokens.push(Token::DoubleQuoted(current_token.clone()));
+                            current_token.clear();
+                        }
+                        _ => unreachable!(),
+                    }
+                }
+                // If the wrong quote is closed, return an error
+                else if quote_stack.contains(&c) {
+                    return Err(Box::new(SyntaxError::UnclosedQuote));
+                }
+                // Otherwise, push the quote onto the stack
+                else {
+                    quote_stack.push(c);
+                }
+            }
+            _ => {
+                debug!("Adding char to token");
+                current_token.push(c);
+            }
+        }
+    }
+
+    Ok(tokens)
+}
+
+pub fn parse(tokens: Vec<Token>) -> Result<RunnableBox, Box<dyn Error>> {
+    let mut in_pipeline = false;
+    let mut in_and_sequence = false;
+    let mut r_type = RedirectType::None;
+
+    let mut current_command = Vec::<Token>::new();
+    let mut commands = Vec::<Box<dyn ShellCommand>>::new();
+    let mut r_cmd: Option<Box<dyn ShellCommand>> = None;
+    let mut sequence_commands = Vec::<Box<dyn Runnable>>::new();
+    let mut final_commands = Vec::<Box<dyn Runnable>>::new();
+
+    for token in tokens {
```

Comment:
For ref (can't share repo as private)

```rust
from typing import Callable, List, Union

from xdsl.dialects.builtin import ModuleOp
from xdsl.ir import Operation

import choco.dialects.choco_ast as ast
from choco.lexer import Lexer, Token, TokenKind


# Tokens that are only binary operators and their descriptions
binary_ops = {
    TokenKind.ASSIGN: "assign",
    TokenKind.AND: "and",
    TokenKind.OR: "or",
    TokenKind.PLUS: "addition",
    TokenKind.MUL: "multiplication",
    TokenKind.DIV: "division",
    TokenKind.MOD: "modulo",
    TokenKind.EQ: "equality",
    TokenKind.NE: "inequality",
    TokenKind.LT: "less than",
    TokenKind.LE: "less than or equal",
    TokenKind.GT: "greater than",
    TokenKind.GE: "greater than or equal",
    TokenKind.IS: "is",
    TokenKind.AND: "and",
    TokenKind.OR: "or",
}


class Parser:
    """
    A Simple ChocoPy Parser

    Parse the given tokens from the lexer and call the xDSL API to create an AST.
    """

    def __init__(self, lexer: Lexer):
        """
        Create a new parser.

        Initialize parser with the corresponding lexer.
        """
        self.lexer = lexer

    def check(self, expected: Union[List[TokenKind], TokenKind]) -> bool:
        """
        Check that the next token is of a given kind. If a list of n TokenKinds
        is given, check that the next n TokenKinds match the next expected
        ones.

        :param expected: The kind of the token we expect or a list of expected
                         token kinds if we look ahead more than one token at
                         a time.
        :returns: True if the next token has the expected token kind, False
                  otherwise.
        """

        if isinstance(expected, list):
            tokens = self.lexer.peek(len(expected))
            assert isinstance(tokens, list), "List of tokens expected"
            return all([tok.kind == type_ for tok, type_ in zip(tokens, expected)])

        token = self.lexer.peek()
        assert isinstance(token, Token), "Single token expected"
        return token.kind == expected

    def match(self, expected: TokenKind) -> Token:
        """
        Match a token by first checking the token kind. In case the token is of
        the expected kind, we consume the token.  If a token with an unexpected
        token kind is encountered, an error is reported by raising an
        exception.

        The exception shows information about the line where the token was expected.

        :param expected: The kind of the token we expect.
        :returns: The consumed token if the next token has the expected token
                  kind, otherwise a parsing error is reported.
        """

        if self.check(expected):
            token = self.lexer.peek()
            assert isinstance(token, Token), "A single token expected"
            self.lexer.consume()
            return token

        token = self.lexer.peek()
        assert isinstance(token, Token), "A single token expected"

        if token.kind == TokenKind.INDENT:
            self.error("Unexpected indentation.")

        self.error_expected(expected)
        assert False, "Unreachable"

    def match_identifier(self) -> Token:
        """
        Match an identifier.

        :return: Operation
        """
        if self.is_literal_type_set():
            self.error("Expected identifier. Found type.")

        return self.match(TokenKind.IDENTIFIER)

    def error(self, msg: str, token: Token | None = None):
        """
        Report a parsing error.

        :param msg: The error message to report.
        :param token: The token that caused the error.
        """
        if token is None:
            peek_token = self.lexer.peek()
            assert isinstance(peek_token, Token), "A single token expected"
            token = peek_token

        line, column = token.location

        print(f"SyntaxError (line {line}, column {column}): {msg}")
        print(f">>>{self.lexer.source_line(token)}")
        print(f">>>{'-' * (column - 1)}^")
        raise Exception("error")

    def error_expected(self, expected: TokenKind, token: Token | None = None):
        """
        Report an error for an expected token kind.

        :param expected: The expected token kind.
        :param token: The token that caused the error.
        """
        self.error(f"token of kind {expected} not found.", token)

    def parse_program(self) -> ModuleOp:
        """
        Parse a ChocoPy program.

        program ::= def_seq stmt_seq EOF

        :returns: The AST of a ChocoPy Program.
        """
        defs = self.parse_def_seq()
        stmts = self.parse_stmt_seq()

        if self.check(TokenKind.DEF):
            self.error("Function declaration after non-declaration statement.")
        if self.check([TokenKind.IDENTIFIER, TokenKind.COLON]):
            self.match(TokenKind.IDENTIFIER)
            self.error("Variable declaration after non-declaration statement.")

        if self.check(TokenKind.CLASS):
            self.error("Class definition not supported.")

        if self.check(TokenKind.GLOBAL):
            self.error("Global declaration outside of function.")

        if self.check(TokenKind.NONLOCAL):
            self.error("Nonlocal declaration outside of function.")

        self.match(TokenKind.EOF)

        return ModuleOp([ast.Program(defs, stmts)])

    def is_def_first_set(self) -> bool:
        """
        Check if the next token is in the first set of a def.
        """
        return (
            self.is_literal_type_set()
            or self.check([TokenKind.IDENTIFIER, TokenKind.COLON])
            or self.check(TokenKind.DEF)
            or self.check(TokenKind.CLASS)
        )

    def parse_def_seq(self) -> List[Operation]:
        """
        Parse a sequence of function and variable definitions.

        def_seq ::= [func_def | var_def]*

        :returns: A list of function and variable definitions.
        """

        defs: List[Operation] = []

        while self.is_def_first_set():
            if self.check(TokenKind.DEF):
                func_def = self.parse_function()
                defs.append(func_def)
            elif self.is_literal_type_set():
                token = self.match_literal_type()
                if self.check(TokenKind.COLON):
                    self.error("Expected identifier. Found type.", token)
                else:
                    self.error("Expected literal. Found type.", token)
            elif self.check([TokenKind.IDENTIFIER, TokenKind.COLON]):
                var_def = self.parse_variable_def()
                defs.append(var_def)
            elif self.check(TokenKind.CLASS):
                self.error("Class definition not supported.")

        return defs

    def parse_function(self) -> Operation:
        """
        Parse a function definition.

         func_def := `def` ID `(` func_params `)` func_return `:` NEWLINE INDENT func_body DEDENT
        func_body := func_var_decls stmt_seq

        :return: Operation
        """
        self.match(TokenKind.DEF)

        function_name = self.match_identifier()

        self.match(TokenKind.LROUNDBRACKET)

        parameters: List[Operation] = self.parse_func_params()
        self.match(TokenKind.RROUNDBRACKET)

        return_type = self.parse_func_return()

        self.match(TokenKind.COLON)

        self.match(TokenKind.NEWLINE)

        if not self.check(TokenKind.INDENT):
            self.error("expected at least one indented statement in function.")

        self.match(TokenKind.INDENT)

        var_decls: List[Operation] = self.parse_func_var_decls()

        if self.check(TokenKind.INDENT):
            self.error("Unexpected indentation.")
        if self.check(TokenKind.DEDENT):
            self.error("expected at least one indented statement in function.")
        if self.check(TokenKind.DEF):
            self.error("Function declaration only allowed at top level.")

        stmt_seq = self.parse_stmt_seq(True)
        if not stmt_seq:
            self.error("expected at least one indented statement in function.")

        if self.check(TokenKind.DEF):
            self.error("Function declaration only allowed at top level.")
        if self.check(TokenKind.GLOBAL):
            self.error("Global declaration after non-declaration statement.")
        if self.check(TokenKind.NONLOCAL):
            self.error("Nonlocal declaration after non-declaration statement.")
        if self.check([TokenKind.IDENTIFIER, TokenKind.COLON]):
            self.match(TokenKind.IDENTIFIER)
            self.error("Variable declaration after non-declaration statement.")

        func_body = var_decls + stmt_seq

        self.match(TokenKind.DEDENT)

        return ast.FuncDef(function_name.value, parameters, return_type, func_body)

    def parse_func_params(self) -> List[Operation]:
        """
        Parse function parameters.

        func_params := [typed_var (',' typed_var)*] | `void`
        typed_var := ID `:` type

        :return: List[Operation]
        """
        if self.check([TokenKind.COLON, TokenKind.NEWLINE]):
            self.error("token of kind TokenKind.RROUNDBRACKET not found.")

        return self.parse_comma_separated_operations(
            lambda: self.check(TokenKind.IDENTIFIER),
            self.parse_typed_var,
            TokenKind.RROUNDBRACKET,
            "input parameter",
        )

    def parse_func_return(self) -> Operation:
        """
        Parse function return type.

        func_return := `->` type | `void`

        :return: Operation
        """
        if self.check(TokenKind.RARROW):
            self.match(TokenKind.RARROW)
            return self.parse_type()

        return ast.TypeName("<None>")

    def parse_func_var_decls(self) -> List[Operation]:
        """
        Parse function variable declarations.

           func_decls := [global_decl | nonlocal_decl | var_def]∗
          global_decl := `global` ID NEWLINE
        nonlocal_decl := `nonlocal` ID NEWLINE
              var_def := typed_var `=` literal NEWLINE

        :return: List[Operation]
        """
        decls: List[Operation] = []

        while True:
            if self.check(TokenKind.GLOBAL):
                self.match(TokenKind.GLOBAL)
                var_name = self.match_identifier()
                self.match(TokenKind.NEWLINE)
                decls.append(ast.GlobalDecl(var_name.value))
            elif self.check(TokenKind.NONLOCAL):
                self.match(TokenKind.NONLOCAL)
                var_name = self.match_identifier()
                self.match(TokenKind.NEWLINE)
                decls.append(ast.NonLocalDecl(var_name.value))
            elif self.is_literal_type_set():
                token = self.match_literal_type()
                if self.check(TokenKind.COLON):
                    self.error("Expected identifier. Found type.", token)
                else:
                    self.error("Expected expression. Found type.", token)
            elif self.check([TokenKind.IDENTIFIER, TokenKind.COLON]):
                decls.append(self.parse_variable_def())
            else:
                break

        return decls

    def parse_variable_def(self) -> Operation:
        """
        Parse a variable definition.

        var_def := typed_var `=` literal NEWLINE

        :return: Operation
        """
        typed_var = self.parse_typed_var()
        self.match(TokenKind.ASSIGN)

        literal = self.parse_literal()

        if literal is None:
            self.error("Expected literal.")

        self.match(TokenKind.NEWLINE)

        return ast.VarDef(typed_var, literal)

    def parse_typed_var(self) -> Operation:
        """
        Parse a typed variable.

        typed_var := ID `:` type

        :return: Operation
        """
        var_name = self.match_identifier()

        self.match(TokenKind.COLON)

        var_type = self.parse_type()

        return ast.TypedVar(var_name.value, var_type)

    def is_literal_type_set(self) -> bool:
        """
        Check if the next token is in the set of a literal (i.e. not list) type.
        """
        return (
            self.check(TokenKind.OBJECT)
            or self.check(TokenKind.INT)
            or self.check(TokenKind.BOOL)
            or self.check(TokenKind.STR)
        )

    def match_literal_type(self) -> Token | None:
        """
        Match a literal type.

        :return: Operation
        """
        if self.check(TokenKind.OBJECT):
            return self.match(TokenKind.OBJECT)
        if self.check(TokenKind.INT):
            return self.match(TokenKind.INT)
        if self.check(TokenKind.BOOL):
            return self.match(TokenKind.BOOL)
        if self.check(TokenKind.STR):
            return self.match(TokenKind.STR)

        return None

    def parse_type(self) -> Operation:
        """
        Parse a type.

             type := type_name | `[` type `]`
        type_name := `object`
                   | `int`
                   | `bool`
                   | `str`

        :return: Operation
        """
        if self.check(TokenKind.LSQUAREBRACKET):
            self.match(TokenKind.LSQUAREBRACKET)
            type = self.parse_type()
            self.match(TokenKind.RSQUAREBRACKET)
            return ast.ListType(type)

        type_name = self.match_literal_type()

        if type_name is None:
            if self.check(TokenKind.IDENTIFIER):
                self.error("Unknown type.")
            else:
                self.error("Excepted type.")
            assert False, "Unreachable"

        return ast.TypeName(type_name.value)

    def is_expr_first_set(self) -> bool:
        """
        Check if the next token is in the first set of an expression.
        """
        return (
            self.check(TokenKind.NOT)
            or self.check(TokenKind.LSQUAREBRACKET)
            or self.check(TokenKind.LROUNDBRACKET)
            or self.check(TokenKind.IDENTIFIER)
            or self.check(TokenKind.MINUS)
            or self.check(TokenKind.NONE)
            or self.check(TokenKind.TRUE)
            or self.check(TokenKind.FALSE)
            or self.check(TokenKind.INTEGER)
            or self.check(TokenKind.STRING)
        )

    def is_expr_NOT_first_set(self) -> bool:
        """
        Check if the next token is in an expression but not first, for error checking.
        """
        for token in binary_ops:
            if self.check(token):
                return True

        return (
            self.check(TokenKind.RROUNDBRACKET)
            or self.check(TokenKind.RSQUAREBRACKET)
            or self.check(TokenKind.ELSE)
            or self.check(TokenKind.COMMA)
        )

    def check_stmt_NOT_first_set(self, token: Token | None = None):
        if self.check(TokenKind.ELIF):
            self.error("Expected expression. Unexpected elif without if.")

        if self.check(TokenKind.ELSE):
            self.error("Expected expression. Unexpected else without if.")

        if self.check(TokenKind.IN):
            if token is not None:
                self.error_expected(TokenKind.FOR, token)
            self.error("Expected expression. Unexpected in outside for loop.")

    def is_stmt_NOT_first_set(self) -> bool:
        """
        Check if the next token is in a statement but not first, for error checking.
        """
        return (
            self.check(TokenKind.ELIF)
            or self.check(TokenKind.IN)
            or self.check(TokenKind.COMMA)
            or self.check(TokenKind.COLON)
            or self.check(TokenKind.RARROW)
            or self.is_literal_type_set()
        )

    def is_stmt_first_set(self) -> bool:
        """
        Check if the next token is in the first set of a statement.
        """
        return (
            # Checks that it is not a variable declaration
            (self.is_expr_first_set() and not self.check([TokenKind.IDENTIFIER, TokenKind.COLON]))
            or self.is_expr_NOT_first_set()
            or self.check(TokenKind.PASS)
            or self.check(TokenKind.RETURN)
            or self.check(TokenKind.IF)
            or self.check(TokenKind.WHILE)
            or self.check(TokenKind.FOR)
            or self.is_stmt_NOT_first_set()
        )

    def parse_stmt_seq(self, requires: bool = False) -> List[Operation]:
        """
        Parse a sequence of statements.

        stmt_seq := stmt stmt_seq | ε

        :param requires: If True, at least one statement is required.
        :return: list of Operations
        """
        stmt_seq: List[Operation] = [self.parse_stmt()] if requires else []
        while self.is_stmt_first_set():
            stmt_op = self.parse_stmt()
            stmt_seq.append(stmt_op)

        return stmt_seq

    def parse_stmt(self) -> Operation:
        """
        Parse a statement.

        stmt := simple_stmt NEWLINE
              | `if` expr `:` block [`elif` expr `:` block]* [`else` `:` block]?
              | `while` expr `:` block
              | `for` ID `in` expr `:` block

        :return: Statement as operation
        """
        if self.check(TokenKind.RROUNDBRACKET):
            self.error("unmatched ')'.")
        if self.check(TokenKind.RSQUAREBRACKET):
            self.error("unmatched ']'.")

        self.check_stmt_NOT_first_set()

        if self.check(TokenKind.IF):
            token = self.match(TokenKind.IF)
            return self.parse_if_stmt_helper(token)

        if self.check(TokenKind.WHILE):
            self.match(TokenKind.WHILE)
            condition = self.parse_expr()
            self.match(TokenKind.COLON)
            block = self.parse_block()
            return ast.While(condition, block)

        if self.check(TokenKind.FOR):
            self.match(TokenKind.FOR)
            var_name = self.match_identifier()
            self.match(TokenKind.IN)
            iterable = self.parse_expr()
            self.match(TokenKind.COLON)
            block = self.parse_block()
            return ast.For(var_name.value, iterable, block)

        first_token = self.lexer.peek()
        assert isinstance(first_token, Token), "A single token expected"

        simple_stmt = self.parse_simple_stmt()

        self.check_stmt_NOT_first_set(first_token)

        if self.check(TokenKind.COLON):
            self.error("Expected end of expression. Unexpected colon.")

        self.match(TokenKind.NEWLINE)
        return simple_stmt

    def parse_if_stmt_helper(self, start_token: Token) -> Operation:
        condition = self.parse_expr()

        if self.check(TokenKind.ELSE):
            self.error("No left-hand side in ternary statement.", start_token)

        self.match(TokenKind.COLON)
        block = self.parse_block()

        if self.check(TokenKind.ELIF):
            token = self.match(TokenKind.ELIF)
            elif_stmt = self.parse_if_stmt_helper(token)
            return ast.If(condition, block, [elif_stmt])

        if self.check(TokenKind.ELSE):
            self.match(TokenKind.ELSE)
            self.match(TokenKind.COLON)
            orelse = self.parse_block()
            return ast.If(condition, block, orelse)

        return ast.If(condition, block, [])

    def parse_block(self) -> List[Operation]:
        """
        Parse a block.

        block := NEWLINE INDENT stmt_seq DEDENT

        :return: list of Operations
        """
        self.match(TokenKind.NEWLINE)

        if not self.check(TokenKind.INDENT):
            self.error("expected at least one indented statement in block.")
        self.match(TokenKind.INDENT)

        if self.check(TokenKind.DEF):
            self.error("Function declaration only allowed at top level.")
        if self.check(TokenKind.CLASS):
            self.error("Class definition not supported.")
        if self.check([TokenKind.IDENTIFIER, TokenKind.COLON]):
            self.error("Variable declaration can only appear at start of function.")
        if self.check(TokenKind.GLOBAL):
            self.error("Global declaration can only appear at start of function.")
        if self.check(TokenKind.NONLOCAL):
            self.error("Nonlocal declaration can only appear at start of function.")

        stmt_seq = self.parse_stmt_seq(True)

        if self.check(TokenKind.DEF):
            self.error("Function declaration only allowed at top level.")
        if self.check(TokenKind.CLASS):
            self.error("Class definition not supported.")
        if self.check([TokenKind.IDENTIFIER, TokenKind.COLON]):
            self.error("Variable declaration can only appear at start of function.")
        if self.check(TokenKind.GLOBAL):
            self.error("Global declaration can only appear at start of function.")
        if self.check(TokenKind.NONLOCAL):
            self.error("Nonlocal declaration can only appear at start of function.")

        self.match(TokenKind.DEDENT)

        return stmt_seq

    def parse_simple_stmt(self) -> Operation:
        """
        Parse a simple statement.

        simple_stmt := `pass`
                     | expr
                     | `return` [expr]?
                     | [expr `=`]+ expr

        :return: Statement as operation
        """
        if self.check(TokenKind.PASS):
            self.match(TokenKind.PASS)
            return ast.Pass()

        if self.check(TokenKind.RETURN):
            self.match(TokenKind.RETURN)

            if self.check(TokenKind.NEWLINE):
                return ast.Return(None)

            expr = self.parse_expr()
            return ast.Return(expr)

        expr = self.parse_expr()

        expr = self.parse_assign_helper(expr)

        if self.check(TokenKind.RROUNDBRACKET):
            self.error("unmatched ')'.")
        if self.check(TokenKind.RSQUAREBRACKET):
            self.error("unmatched ']'.")

        if self.check(TokenKind.COMMA):
            self.error("Comma outside list or function call.")
        if self.check(TokenKind.RARROW):
            self.error("Right arrow outside function definition.")

        return expr

    def parse_assign_helper(self, expr: Operation) -> Operation:
        if not self.check(TokenKind.ASSIGN):
            return expr

        self.match(TokenKind.ASSIGN)
        value = self.parse_expr()
        return ast.Assign(expr, self.parse_assign_helper(value))

    def parse_expr(self, precedence: int = 0) -> Operation:
        """
        Parse an expression.

        expr := cexpr
              | `not` expr
              | expr [`and` | `or`] expr
              | expr `if` expr `else` expr

        :param precedence: Precedence of the current expression.
        :return: Operation
        """
        for token, desc in binary_ops.items():
            if self.check(token):
                self.error(f"No left-hand side in {desc} statement.")

        if self.check(TokenKind.NOT) and precedence <= 4:
            self.match(TokenKind.NOT)
            expr = ast.UnaryExpr("not", self.parse_expr(4))
        else:
            expr = self.parse_cexpr(precedence)

        if precedence < 3:
            while self.check(TokenKind.AND):
                self.match(TokenKind.AND)
                right = self.parse_expr(3)
                expr = ast.BinaryExpr("and", expr, right)

        if precedence < 2:
            while self.check(TokenKind.OR):
                self.match(TokenKind.OR)
                right = self.parse_expr(2)
                expr = ast.BinaryExpr("or", expr, right)

        if precedence <= 1:
            if self.check(TokenKind.IF):
                self.match(TokenKind.IF)
                cond = self.parse_expr(1)

                if self.check(TokenKind.ELIF):
                    self.error("Ternary expression does not support elif.")

                self.match(TokenKind.ELSE)
                alt = self.parse_expr(precedence)
                expr = ast.IfExpr(cond, expr, alt)

        return expr

    def parse_cexpr(self, precedence: int = 0) -> Operation:
        """
        Parse a conditional expression.

        cexpr := ID
               | literal
               | `[` expr_list `]`
               | `(` expr `)`
               | index_expr
               | ID `(` expr_list `)`
               | cexpr bin_op cexpr
               | `-` cexpr

        :param precedence: Precedence of the current expression.
        :return: Operation
        """
        if self.check(TokenKind.LROUNDBRACKET):
            self.match(TokenKind.LROUNDBRACKET)
            expr = self.parse_expr()
            self.match(TokenKind.RROUNDBRACKET)
        else:
            expr = self.parse_literal()

        if expr is None:
            if self.check(TokenKind.IDENTIFIER):
                identifier = self.match(TokenKind.IDENTIFIER)

                if self.check(TokenKind.LROUNDBRACKET):
                    self.match(TokenKind.LROUNDBRACKET)
                    args = self.parse_expr_list(end=TokenKind.RROUNDBRACKET)

                    if self.check(TokenKind.COLON):
                        self.error("Parameter declaration not allowed in function call.")

                    self.match(TokenKind.RROUNDBRACKET)
                    expr = ast.CallExpr(identifier.value, args)
                else:
                    expr = ast.ExprName(identifier.value)
            elif self.check(TokenKind.LSQUAREBRACKET):
                self.match(TokenKind.LSQUAREBRACKET)
                elements = self.parse_expr_list(end=TokenKind.RSQUAREBRACKET)
                self.match(TokenKind.RSQUAREBRACKET)
                expr = ast.ListExpr(elements)
            elif precedence <= 8:
                if self.check(TokenKind.MINUS):
                    self.match(TokenKind.MINUS)
                    right = self.parse_cexpr(8)
                    expr = ast.UnaryExpr("-", right)

        if expr is None:
            self.error("Expected expression.")
            assert False, "Unreachable"

        if precedence < 9:
            while self.check(TokenKind.LSQUAREBRACKET):
                self.match(TokenKind.LSQUAREBRACKET)
                index = self.parse_expr()
                self.match(TokenKind.RSQUAREBRACKET)
                expr = ast.IndexExpr(expr, index)

        if precedence < 7:
            while (
                self.check(TokenKind.MUL) or self.check(TokenKind.DIV) or self.check(TokenKind.MOD)
            ):
                if self.check(TokenKind.MUL):
                    token, op = self.match(TokenKind.MUL), "*"
                elif self.check(TokenKind.DIV):
                    token, op = self.match(TokenKind.DIV), "/"
                elif self.check(TokenKind.MOD):
                    token, op = self.match(TokenKind.MOD), "%"

                assert token is not None and op is not None, "Invalid token or op 7"

                right = self.parse_cexpr(7)
                expr = ast.BinaryExpr(token.value, expr, right)

        if precedence < 6:
            while self.check(TokenKind.PLUS) or self.check(TokenKind.MINUS):
                if self.check(TokenKind.PLUS):
                    token, op = self.match(TokenKind.PLUS), "+"
                elif self.check(TokenKind.MINUS):
                    token, op = self.match(TokenKind.MINUS), "-"

                assert token is not None and op is not None, "Invalid token or op 6"

                right = self.parse_cexpr(6)
                expr = ast.BinaryExpr(token.value, expr, right)

        if precedence < 5:
            token, op = None, None
            if self.check(TokenKind.EQ):
                token, op = self.match(TokenKind.EQ), "=="
            elif self.check(TokenKind.NE):
                token, op = self.match(TokenKind.NE), "!="
            elif self.check(TokenKind.LT):
                token, op = self.match(TokenKind.LT), "<"
            elif self.check(TokenKind.LE):
                token, op = self.match(TokenKind.LE), "<="
            elif self.check(TokenKind.GT):
                token, op = self.match(TokenKind.GT), ">"
            elif self.check(TokenKind.GE):
                token, op = self.match(TokenKind.GE), ">="
            elif self.check(TokenKind.IS):
                token, op = self.match(TokenKind.IS), "is"

            if token is not None and op is not None:
                right = self.parse_cexpr(5)
                expr = ast.BinaryExpr(op, expr, right)
        elif precedence == 5:
            if (
                self.check(TokenKind.EQ)
                or self.check(TokenKind.NE)
                or self.check(TokenKind.LT)
                or self.check(TokenKind.LE)
                or self.check(TokenKind.GT)
                or self.check(TokenKind.GE)
                or self.check(TokenKind.IS)
            ):
                self.error("Comparison operators are not associative.")

        return expr

    def parse_expr_list(self, end: TokenKind) -> List[Operation]:
        """
        Parse a list of expressions.

        expr_list := [expr [`,` expr]*]?

        :return: List of expressions as operations
        """

        return self.parse_comma_separated_operations(
            self.is_expr_first_set, self.parse_expr, end, "expression"
        )

    def parse_comma_separated_operations(
        self,
        first_set: Callable[[], bool],
        get_operation: Callable[[], Operation],
        end: TokenKind,
        operation_desc: str,
    ) -> List[Operation]:
        """
        Parse a list of comma separated operations.

        :param get_operation: A function that returns the next operation.
        :param end: The token that ends the list of operations.
        :return: List of operations
        """
        if self.check(end):
            return []

        if self.check(TokenKind.NEWLINE):
            self.error_expected(end)

        if self.check(TokenKind.COMMA):
            self.error(f"Expected {operation_desc}.")

        operation_list = [get_operation()]

        while self.check(TokenKind.COMMA):
            self.match(TokenKind.COMMA)
            operation_list.append(get_operation())

        if first_set():
            # NOTE: It would be better to use operation description
            # self.error(f"{operation_desc} found, but comma expected.")
            self.error("expression found, but comma expected.")

        if not self.check(end):
            if self.check(TokenKind.NEWLINE):
                self.error_expected(end)

            token = self.lexer.peek()
            assert isinstance(token, Token), "A single token expected"
            first_token = token

            while not self.check(end):
                self.match(token.kind)

                if first_set():
                    self.error_expected(TokenKind.COMMA, first_token)

                if self.check(end) or self.check(TokenKind.NEWLINE):
                    break

                token = self.lexer.peek()
                assert isinstance(token, Token), "A single token expected"

            self.error_expected(end, first_token)

        return operation_list

    def parse_literal(self) -> Operation | None:
        """
        Parse a literal.

        literal := `None`
                 | `True`
                 | `False`
                 | INTEGER
                 | STRING

        :return: Operation
        """
        if self.check(TokenKind.NONE):
            self.match(TokenKind.NONE)
            return ast.Literal(None)

        if self.check(TokenKind.TRUE):
            self.match(TokenKind.TRUE)
            return ast.Literal(True)

        if self.check(TokenKind.FALSE):
            self.match(TokenKind.FALSE)
            return ast.Literal(False)

        if self.check(TokenKind.INTEGER):
            integer = self.match(TokenKind.INTEGER)
            return ast.Literal(integer.value)

        if self.check(TokenKind.STRING):
            string = self.match(TokenKind.STRING)
            return ast.Literal(string.value)

        if self.is_literal_type_set():
            self.error("Expected expression. Found type.")

        return None
```

Reply:
Beautiful stuff :cry: